# Script

In [0]:
!unzip drive*

#Imports

In [0]:
import pandas as pd
import glob
import codecs
import re
import matplotlib.pyplot as plt
import nltk
from nltk.stem import PorterStemmer
from nltk.corpus import stopwords
import numpy as np
from sklearn.metrics import f1_score

#Configurations

In [0]:
nltk.download('rslp')
nltk.download('stopwords')
nltk.download('punkt')

[nltk_data] Downloading package rslp to /root/nltk_data...
[nltk_data]   Unzipping stemmers/rslp.zip.
[nltk_data] Downloading package stopwords to /root/nltk_data...
[nltk_data]   Unzipping corpora/stopwords.zip.
[nltk_data] Downloading package punkt to /root/nltk_data...
[nltk_data]   Unzipping tokenizers/punkt.zip.


True

##Loads

In [0]:
files_true = []

for file in sorted(glob.glob("/content/Artigos Originais/*.txt")):
  try:
    print(file)
    tfh = codecs.open(file, "rb", encoding="utf-8")
    files_true.append(tfh.readlines())
  except:
    print("Erro ao ler tentando outro encoding")
    tfh = codecs.open(file, "rb", encoding="ISO-8859-1")
    files_true.append(tfh.readlines())

/content/Artigos Originais/Artigo 1.txt
Erro ao ler tentando outro encoding
/content/Artigos Originais/Artigo 11.txt
/content/Artigos Originais/Artigo 12.txt
/content/Artigos Originais/Artigo 13.txt
/content/Artigos Originais/Artigo 14.txt
/content/Artigos Originais/Artigo 16.txt
/content/Artigos Originais/Artigo 2.txt
/content/Artigos Originais/Artigo 3.txt
/content/Artigos Originais/Artigo 4.txt
/content/Artigos Originais/Artigo 5.txt
/content/Artigos Originais/Artigo 6.txt
/content/Artigos Originais/Artigo 9.txt


In [0]:
files_pred = []

for file in sorted(glob.glob("/content/Artigos Modificados/*.txt")):
  try:
    print(file)
    tfh = codecs.open(file, "rb", encoding="utf-8")
    files_pred.append(tfh.readlines())
  except:
    print("Erro ao ler tentando outro encoding")
    tfh = codecs.open(file, "rb", encoding="ISO-8859-1")
    files_pred.append(tfh.readlines())

/content/Artigos Modificados/Artigo 1 - Alterado.txt
/content/Artigos Modificados/Artigo 11 - Alterado.txt
/content/Artigos Modificados/Artigo 12 - Alterado.txt
/content/Artigos Modificados/Artigo 13 - Alterado.txt
/content/Artigos Modificados/Artigo 14 - Alterado.txt
/content/Artigos Modificados/Artigo 16 - Alterado.txt
/content/Artigos Modificados/Artigo 2 - Alterado.txt
/content/Artigos Modificados/Artigo 3 - Alterado.txt
/content/Artigos Modificados/Artigo 4 - Alterado.txt
/content/Artigos Modificados/Artigo 5 - Alterado.txt
/content/Artigos Modificados/Artigo 6 - Alterado.txt
/content/Artigos Modificados/Artigo 9 - Alterado.txt


#Functions

In [0]:

def clean(files):
  clean_files = []
  all_doc = files
  #all_doc = ' '.join(files)
  clean_files.append(re.sub('\s+', ' ', all_doc))
  one_doc = ' '.join(clean_files)
  tokens = nltk.word_tokenize(one_doc)
  stemming = PorterStemmer()
  stem = [stemming.stem(word) for word in tokens]
  stops = set(stopwords.words("english"))      
  stops_eng = set(stopwords.words("english"))    
  filtered_words = [word for word in stem if word not in stops]
  filtered_words_1 = [word for word in filtered_words if word not in stops_eng]
  filtered_words_only =list(filter(lambda x:x, map(lambda x:re.sub(r'[^A-Za-z]', '', x), filtered_words_1)))
  filtered_words_only = [i for i in filtered_words_only if len(i) > 1]
  return filtered_words_only

def is_number(s):
    """

    Verifica se a variavel é um número
    
    Parameters
    ----------
    image : int
        The first parameter.

    Returns
    -------
    bool
        Verdadeiro caso o cast ocorra. Falso caso contrário.
    """ 
    try:
        float(s)
        return True
    except ValueError:
        return False

def sugg_napas(doc):
  seed_words = clean(doc)
  corrected = []
  for i in seed_words:
    if not natas.is_correctly_spelled(i) and not is_number(i) and i:
      sug = natas.ocr_correct_words([i])[0]
      if sug:
        corrected.append(sug[0])
      else:
        corrected.append(i)
    elif i:
      corrected.append(i)
  return corrected

#Main Statistics Napas

##Loads

In [0]:
files_true = []

for file in sorted(glob.glob("/content/Artigos Originais/*.txt")):
  try:
    print(file)
    tfh = codecs.open(file, "rb", encoding="utf-8")
    files_true.append(tfh.readlines())
  except:
    print("Erro ao ler tentando outro encoding")
    tfh = codecs.open(file, "rb", encoding="ISO-8859-1")
    files_true.append(tfh.readlines())

/content/Artigos Originais/Artigo 1.txt
Erro ao ler tentando outro encoding
/content/Artigos Originais/Artigo 11.txt
/content/Artigos Originais/Artigo 12.txt
/content/Artigos Originais/Artigo 13.txt
/content/Artigos Originais/Artigo 14.txt
/content/Artigos Originais/Artigo 16.txt
/content/Artigos Originais/Artigo 2.txt
/content/Artigos Originais/Artigo 3.txt
/content/Artigos Originais/Artigo 4.txt
/content/Artigos Originais/Artigo 5.txt
/content/Artigos Originais/Artigo 6.txt
/content/Artigos Originais/Artigo 9.txt


In [0]:
files_pred = []

for file in sorted(glob.glob("/content/Artigos Modificados/*.txt")):
  try:
    print(file)
    tfh = codecs.open(file, "rb", encoding="utf-8")
    files_pred.append(tfh.readlines())
  except:
    print("Erro ao ler tentando outro encoding")
    tfh = codecs.open(file, "rb", encoding="ISO-8859-1")
    files_pred.append(tfh.readlines())

/content/Artigos Modificados/Artigo 1 - Alterado.txt
/content/Artigos Modificados/Artigo 11 - Alterado.txt
/content/Artigos Modificados/Artigo 12 - Alterado.txt
/content/Artigos Modificados/Artigo 13 - Alterado.txt
/content/Artigos Modificados/Artigo 14 - Alterado.txt
/content/Artigos Modificados/Artigo 16 - Alterado.txt
/content/Artigos Modificados/Artigo 2 - Alterado.txt
/content/Artigos Modificados/Artigo 3 - Alterado.txt
/content/Artigos Modificados/Artigo 4 - Alterado.txt
/content/Artigos Modificados/Artigo 5 - Alterado.txt
/content/Artigos Modificados/Artigo 6 - Alterado.txt
/content/Artigos Modificados/Artigo 9 - Alterado.txt


##Napas

In [0]:
#%%shell 
#!pip3 install natas
#!python3 -m natas.download
#!python3 - m spacy download en_core_web_md
!python3 - m spacy download pt_core_news_sm


In [0]:
import spacy
import natas
nlp = spacy.load('pt')

In [0]:
f_score = []
for y_true, y_predicted in list(zip(files_true, files_pred)):
  y_true =  clean(y_true)
  y_predicted = clean(y_predicted)
  y_predicted = sugg_napas(y_predicted)
  if len(y_true) < len(y_predicted):
    y_predicted = y_predicted[:len(y_true)]
  else:
    y_true = y_true[:len(y_predicted)]
 
  f_score.append(f1_score(y_true, y_predicted, average='micro'))

In [0]:
f_score_np = np.asarray(f_score)

In [0]:
mean = np.mean(f_score_np)
median = np.median(f_score_np)
std = np.std(f_score_np)

In [0]:
print("F1 Score de cada documento")
print("Média {}".format(round(mean, 3)))
print("Mediana {}".format(round(median, 3)))
print("Desvio Padrão {}".format(round(std, 3)))

F1 Score de cada documento
Média 0.259
Mediana 0.167
Desvio Padrãão 0.247


F1 Score de cada documento
* Média 0.259
* Mediana 0.167
* Desvio Padrão 0.247

#Main Statistics CCC

##Loads

In [0]:
files_true = []

for file in sorted(glob.glob("/content/Artigos Originais/*.txt")):
  try:
    print(file)
    tfh = codecs.open(file, "rb", encoding="utf-8")
    files_true.append(tfh.readlines())
  except:
    print("Erro ao ler tentando outro encoding")
    tfh = codecs.open(file, "rb", encoding="ISO-8859-1")
    files_true.append(tfh.readlines())

/content/Artigos Originais/Artigo 1.txt
Erro ao ler tentando outro encoding
/content/Artigos Originais/Artigo 11.txt
/content/Artigos Originais/Artigo 12.txt
/content/Artigos Originais/Artigo 13.txt
/content/Artigos Originais/Artigo 14.txt
/content/Artigos Originais/Artigo 16.txt
/content/Artigos Originais/Artigo 2.txt
/content/Artigos Originais/Artigo 3.txt
/content/Artigos Originais/Artigo 4.txt
/content/Artigos Originais/Artigo 5.txt
/content/Artigos Originais/Artigo 6.txt
/content/Artigos Originais/Artigo 9.txt


In [0]:
files_pred = []

for file in sorted(glob.glob("/content/Artigos Modificados/*.txt")):
  try:
    print(file)
    tfh = codecs.open(file, "rb", encoding="utf-8")
    files_pred.append(tfh.readlines())
  except:
    print("Erro ao ler tentando outro encoding")
    tfh = codecs.open(file, "rb", encoding="ISO-8859-1")
    files_pred.append(tfh.readlines())

/content/Artigos Modificados/Artigo 1 - Alterado.txt
/content/Artigos Modificados/Artigo 11 - Alterado.txt
/content/Artigos Modificados/Artigo 12 - Alterado.txt
/content/Artigos Modificados/Artigo 13 - Alterado.txt
/content/Artigos Modificados/Artigo 14 - Alterado.txt
/content/Artigos Modificados/Artigo 16 - Alterado.txt
/content/Artigos Modificados/Artigo 2 - Alterado.txt
/content/Artigos Modificados/Artigo 3 - Alterado.txt
/content/Artigos Modificados/Artigo 4 - Alterado.txt
/content/Artigos Modificados/Artigo 5 - Alterado.txt
/content/Artigos Modificados/Artigo 6 - Alterado.txt
/content/Artigos Modificados/Artigo 9 - Alterado.txt


##Modified CCC 

Não teve os dados para que pudesse ser feito a análise

In [0]:
def load_text(file_path):
    text = []
    with codecs.open(file_path, encoding="utf-8-sig") as f:
        for line in f:
            text.append(line)
    return ' '.join(text)

In [0]:
files_pred= load_text("/content/Rapunzel_250_test_albert_corredtec.txt").split('\n')
files_true = load_text("/content/Rapunzel_250_original.txt").split('\n')

In [0]:
f_score = []
for y_true, y_predicted in list(zip(files_true, files_pred)):
  y_true =  clean(y_true)
  y_predicted = clean(y_predicted)
  print("++++++++++")
  print(y_true)
  print(y_predicted)
  if len(y_true) < len(y_predicted):
    y_predicted = y_predicted[:len(y_true)]
  else:
    y_true = y_true[:len(y_predicted)]
 
  f_score.append(f1_score(y_true, y_predicted, average='micro'))

++++++++++
['onc', 'man', 'woman', 'long', 'vain', 'wish', 'child']
['onc', 'man', 'woman', 'long', 'vain', 'wish', 'child']
++++++++++
['At', 'length', 'woman', 'hope', 'god', 'wa', 'grant', 'desir']
['At', 'length', 'woman', 'hope', 'god', 'wa', 'grant', 'desir']
++++++++++
['peopl', 'littl', 'window', 'back', 'hous', 'splendid', 'garden', 'could', 'seen', 'wa', 'full', 'beauti', 'flower', 'herb']
['peopl', 'littl', 'window', 'back', 'hous', 'splendid', 'garden', 'could', 'seen', 'wa', 'full', 'beauti', 'flower', 'herb']
++++++++++
['It', 'wa', 'howev', 'surround', 'high', 'wall', 'one', 'dare', 'go', 'becaus', 'belong', 'enchantress', 'great', 'power', 'wa', 'dread', 'world']
['It', 'wa', 'howev', 'surround', 'high', 'call', 'one', 'dare', 'go', 'becaus', 'belong', 'enchantress', 'great', 'power', 'wa', 'dread', 'world']
++++++++++
['one', 'day', 'woman', 'wa', 'stand', 'thi', 'window', 'look', 'garden', 'saw', 'bed', 'wa', 'plant', 'beauti', 'flower', 'look', 'fresh', 'green', 'lon

In [0]:
f_score_np = np.asarray(f_score)

In [0]:
mean = np.mean(f_score_np)
median = np.median(f_score_np)
std = np.std(f_score_np)

In [0]:
print("F1 Score de cada documento")
print("Média {}".format(round(mean, 3)))
print("Mediana {}".format(round(median, 3)))
print("Desvio Padrão {}".format(round(std, 3)))

F1 Score de cada documento
Média 0.909
Mediana 1.0
Desvio Padrão 0.19
